### DISCUSSION TITLE
Android example

### DISCUSSION THREAD

**Discussion by Malachiasz** [classgraph/classgraph#567] [Ideas]
Please provide a sample showing what is possible on Android.

**Comment by lukehutch** [comment]
Android does not use classfiles at time of deployment, therefore ClassGraph cannot be used. However, ClassGraph can be used at build time in Android projects, before the compiled code is converted from Java bytecodes to Dalvik bytecodes. This is documented here:

https://github.com/classgraph/classgraph/wiki/Build-Time-Scanning

**Comment by Malachiasz** [comment]
I have found it after asking the question here. The sample did not compile I had to modify it:

    afterEvaluate { project ->
        // Get some task references
        def compileTask = project.tasks.compileDebugJavaWithJavac
        def mergeTask = project.tasks.mergeDebugResources
    
        compileTask.doLast {
            // [Set your parameters here]
            def packageToScan = "my.project"
            def scanResultFilename = "scan_result.txt"
    
            // Create a temporary raw resource directory for saving the scan result to
            def scanResultDir = compileTask.temporaryDir.toString() + "/res/raw/"
            new File(scanResultDir).mkdirs()
            def scanResultPath = scanResultDir + scanResultFilename
    
            // Get location where .class files were compiled to
            def classPackageRoot = compileTask.destinationDir
    
            println "Scanning " + classPackageRoot
            new io.github.classgraph.ClassGraph()
            // [Configure your ClassGraph instance here]
                    .overrideClasspath(classPackageRoot)
                    .whitelistPackages(packageToScan)
                    .enableAllInfo()
            // .verbose()
                    .scan()
                    .withCloseable { scanResult ->
                        // [Get the scan results you're interested in here]
                        def resultList = scanResult.getAllClasses()
                        // Write the results to the output file
                        def outputFile = new File(scanResultPath)
                        outputFile.withWriter { writer ->
                            resultList.each { ci ->
                                // Write name of annotated class to output file
                                writer.println ci.getName()
                            }
                        }
                        println "Wrote scan result to " + outputFile +" ; size = " + outputFile.length()
                    }
    
            // Run AAPT2 to convert the raw resource file into a .flat file
            def aapt2Path = "/Users/user/Library/Android/sdk/build-tools/30.0.0/aapt2" // change it to your path
            def flatFileOutputPrefix = mergeTask.outputDir.toString() + "/"
            def cmd = [aapt2Path, "compile", "-o", flatFileOutputPrefix, scanResultPath]
            println "Executing: " + cmd.join(" ")
            def exec = cmd.execute()
            exec.in.eachLine { line -> println line }
            exec.err.eachLine { line -> System.err.println "ERROR: " + line }
            exec.waitFor()
            println "Wrote compiled resource as a .flat file to " + flatFileOutputPrefix
        }
    }

That's why it's better to have a sample on github with gradle version java version and everything else specified. 

Also I see it cannot access files not in my sources.

**Comment by lukehutch** [comment]
It used to work, so probably Gradle or Android changed, or something... and the same exact bitrot can happen just as easily with sample projects. I don't actively work on Android anymore, as of several years ago. You are welcome to set up your own example project on GitHub for people to discover. But honestly there seems to be very little demand for scanning classes in Android projects -- I get about one question per year about this.

I can update the build-time scanning page though. Thanks for the working example!

**Reply by Malachiasz** [reply]
The same cannot happen with sample project because gradle file of the sample project defines all these variables.

**Reply by lukehutch** [reply]
Yes, I know what you mean, but let's say I wrote this to be compatible with JDK 7 and an old version of Gradle, and you want to use it on JDK 11 and a new version of Gradle that is somehow not compatible, you would still have to change the code to actually use the example. I really don't have the bandwidth to write and maintain example code for every different scenario that people want to run ClassGraph under. ClassGraph supports far too many different runtime scenarios to make that a possibility: https://github.com/classgraph/classgraph/wiki/Classpath-Specification-Mechanisms

But if you would like to see this as an example project, please do publish this as a GitHub project. People will be able to find it with Google search.

**Reply by lukehutch** [reply]
And actually I just compared your changes to the original. The changes you made are changes that anybody would have to make to adapt the example to their specific scanning needs (e.g. scanning all classes rather than classes with a given annotation), maybe with the exception of the aapt2Path change. Did the code needed to find aapt2Path not work? It is supposed to work generically, for all Android builds.

![image](https://user-images.githubusercontent.com/811305/136289191-602df521-1653-4af5-a00c-1f0351b2c068.png)

Also I don't know why you changed `.acceptPackages` to `.whitelistPackages` -- `.whitelistPackages` is deprecated, and `.acceptPackages` is correct.

**Reply by HashirLabs** [reply]
unfortunately, i am getting error on below line
```

def exec = cmd.execute()

ERROR: task ':app:mergeDebugResources' property 'outputDir'/: error: The device does not recognize the command. (22).
```


Below is the lower part of the gradle i am using, 

```
 // Run AAPT2 to convert the raw resource file into a .flat file
        def aapt2Path = "D:/IT/Android/sdk/build-tools/33.0.1/aapt2" // change it to your path
//        def aapt2Path = mergeTask.buildTools.getPath(com.android.sdklib.BuildToolInfo.PathId.AAPT2)
        def flatFileOutputPrefix = mergeTask.outputDir.toString() + "/"
        def cmd = [aapt2Path, "compile", "-o", flatFileOutputPrefix, scanResultPath]
        println "Executing: " + cmd.join(" ")
        def exec = cmd.execute()
        exec.in.eachLine { line -> println line }
        exec.err.eachLine { line -> System.err.println "ERROR: " + line }
        exec.waitFor()
        println "Wrote compiled resource as a .flat file to " + flatFileOutputPrefix
```

any help?

**Reply by lukehutch** [reply]
Sorry, that sounds like a Gradle question, and I have no idea. Maybe try asking about this error message on a Gradle forum?

**Reply by thc202** [reply]
You most likely need to do `mergeTask.outputDir.get().toString()`, it seems that now the `outputDir` is a [`Property`](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Property.html) so you need to get its value before converting to string.

**Reply by HashirLabs** [reply]
Thanks, this worked now

I also want to add that after this i was facing invalid path issue on windows machine, i had to update below line which fixed it and now its working perfectly. 

old
`def scanResultDir = compileTask.temporaryDir.toString() + "/res/raw/"`

New
`def scanResultDir = compileTask.temporaryDir.toString() + "\\res\\raw\\"`

Now, i am wondering if there is a way to provide the aapt path via gradle constants rather than hardcoded path, as this path gets updated as and when we update android build tool

something like def aapt2Path = mergeTask.buildTools.getPath(com.android.sdklib.BuildToolInfo.PathId.AAPT2) which does not work at the moment

instead of def aapt2Path = "D:/IT/Android/sdk/build-tools/33.0.1/aapt2" /

many thanks for the help anywayz

**Reply by lukehutch** [reply]
@HashirLabs thanks, I updated the docs to use `File.separator`.

I don't do any low-level Android build stuff anymore, so I can't help with the aapt2 path, but if you figure it out, let me know, and I'll update the docs again. Thanks!

**Reply by lukehutch** [reply]
@thc202 Thanks for the fix, I updated the wiki with your suggested change

**Comment by HashirLabs** [comment]
@lukehutch 

i have updated my gradle script to pick the aapt2 dynamically based on user's sdk directory, now the manual path setup is not required, below is tested and found working, you can update the wiki accordingly


        def aapt2Path = android.getSdkDirectory().toPath().resolve("build-tools").resolve(android.buildToolsVersion).resolve("aapt2")

**Comment by lukehutch** [comment]
@HashirLabs Nice work! Thanks, I updated the wiki.
